### ISSUE TITLE
CustomXMLPart can't sync

### ISSUE THREAD

**Issue by YbJerry** [OfficeDev/office-js#2937]
# Provide required information needed to triage your issue
I'am trying to use Common Office API to manipulate the Bibliography in the docx file. Because I think Bibiliography will be store in the customXml. So I try to use API about CustomXml in Common Office API to do that thing. At first, I try to add a Bibliography in a new document, it works. But later I find I can't change or add a source in the Bibliography. That's strange.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC desktop
* Host [Excel, Word, PowerPoint, etc.]: Word
* Office version number:  Microsoft 365MSO (Version 2209 Build 16.0.15629.20200) 64 bit
* Operating System: Windows
* Browser (if using Office on the web): ______
 
## Expected behavior
When I use customXmlNode.setXmlAsync, Word should sync this change in the Bibliography.

## Current behavior
I can add a new Bibliography use customXmlPart.addAsync. But later I can't change it.

## Steps to reproduce
1. Open a new blank document, and run this code in the Script Lab
```
async function initBib() {
  Office.context.document.customXmlParts.addAsync(
    `<b:Sources xmlns:b="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" xmlns="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" SelectedStyle="\IEEE2006OfficeOnline.xsl" StyleName="IEEE" Version="2006">
      <b:Source>
          <b:Tag>tes</b:Tag>
          <b:SourceType>Misc</b:SourceType>
          <b:Guid>{C8C15628-8E0D-4E7F-9C13-2B76253B234C}</b:Guid>
          <b:Author>
              <b:Author>
                  <b:NameList>
                      <b:Person>
                          <b:Last>test</b:Last>
                      </b:Person>
                  </b:NameList>
              </b:Author>
          </b:Author>
          <b:LCID>zh-CN</b:LCID>
          <b:RefOrder>1</b:RefOrder>
      </b:Source>
  </b:Sources>`,
    function(res) {
      console.log(res);
    }
  );
}
```
You will find you actually add a Citation in the Word.
2. But when you run following code, you will find there is nothing changed
```
async function addBib() {
  Office.context.document.customXmlParts.getByNamespaceAsync(
    "http://schemas.openxmlformats.org/officeDocument/2006/bibliography",
    function(res) {
      if (res.value.length > 0) {
        res.value[0].getNodesAsync("/*[local-name()='Sources']", (res) => {
          console.log(res);
          if (res.value.length > 0) {
            res.value[0].setXmlAsync(
              '<b:Sources xmlns:b="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" xmlns="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" SelectedStyle="\\IEEE2006OfficeOnline.xsl" StyleName="IEEE" Version="2006"></b:Sources>',
              (res) => {
                console.log(res);
              }
            );
          }
        });
      }
    }
  );
}
```
3. But when you look at the Bibliography using following code, you will find that you have changed the CustomXml. But OfficeJs won't sync it.
```
async function getBib() {
  Office.context.document.customXmlParts.getByNamespaceAsync(
    "http://schemas.openxmlformats.org/officeDocument/2006/bibliography",
    function(res) {
      if (res.value.length > 0) {
        res.value[0].getNodesAsync("/*[local-name()='Sources']", (res) => {
          console.log(res);
          if (res.value.length > 0) {
            res.value[0].getXmlAsync((res) => {
              console.log(res);
          }
        });
      }
    }
  );
}
```
There is the output:
```
{value: "<b:Sources xmlns:b="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" xmlns="http://schemas.openxmlformats.org/officeDocument/2006/bibliography" SelectedStyle="\IEEE2006OfficeOnline.xsl" StyleName="IEEE" Version="2006"></b:Sources>", status: "succeeded"}
```

## Link to live example(s)
1. https://gist.github.com/YbJerry/3bcacf69a4caa414bbe9a1c0bf751934

## Context
I'm trying to write a add-in which can directly add Bibliography into OOXML. Because I found OOXML's document has this part and Word can support this function.

Thank you for taking the time to report an issue. Our triage team will respond to you in less than 72 hours. Normally, response time is <10 hours Monday through Friday. We do not triage on weekends.

**Comment by Searion** [comment]
Hi @YbJerry , I try to repro it and I ran initBib() in ScriptLab or run the example but I did not see anything change. Could you please provide more information about it?

**Comment by YbJerry** [comment]
Yes, when you open a new blank document and run initBib. You will find there is a new citation in the Word.

![图片](https://user-images.githubusercontent.com/34412324/199907193-8010cabe-6a3b-416a-a2d3-0d52e0e9db83.png)

But before this step, please don't do anything else. Even if you just click anything in Citation, this code will fail.

**Comment by YbJerry** [comment]
And if you run addBib(it's actually clear all the Citations, sorry, I forget change its name) and run getBib. You will find it has changed in OfficeJS.
![图片](https://user-images.githubusercontent.com/34412324/199908156-01ca3b78-738b-4a33-842f-41f54b77f542.png)
But Word seems not to sync this change and it still has the citation inserted before. 
![图片](https://user-images.githubusercontent.com/34412324/199908472-16b1911a-1164-4719-94aa-5547d886e2c3.png)

**Comment by YbJerry** [comment]
@Searion

**Comment by jseddon2** [comment]
I have seen the same behaviour as @YbJerry.  I was trying this approach as a potential workaround to there not being a bibliography API (issue #3402) but unfortunately it doesn't work once you open the manage sources dialog (or use the insert citations drop-down).  Are there any workarounds available in the absence of the bibliography API?

**Comment by YbJerry** [comment]
> I have seen the same behaviour as @YbJerry. I was trying this approach as a potential workaround to there not being a bibliography API (issue #3402) but unfortunately it doesn't work once you open the manage sources dialog (or use the insert citations drop-down). Are there any workarounds available in the absence of the bibliography API?

Hi @jseddon2, as Microsoft wouldn't implement bibliography API recently, I will show you my solution. If you want to write a plugin of bibliography, I recommend you to use [CSL](https://citationstyles.org/) and refer these two projects [JabRef-Word-Addin](https://github.com/JabRef/JabRef-Word-Addin) and [citeproc-js](https://github.com/Juris-M/citeproc-js). Citeproc is a tool for managing [CSL](https://citationstyles.org/) and citeproc-js is its JavaScript implement. This library is old and if you want to use TypeScript, you should import [citeproc-types](https://www.npmjs.com/package/citeproc-types?activeTab=readme). And JabRef showed a way to save all the bibliography in the Control.

Maybe you are a Rust programmer and want to use WASM to do same thing, but do not to use [citeproc-rs](https://github.com/zotero/citeproc-rs). I have tried Rust implement, and it has bad support to some CSL styles such as Chinese bibliography style. But you can read its [document](https://github.com/zotero/citeproc-rs/tree/master/crates/wasm) to understand how citeproc works. Citeproc-rs really have a better document than citeproc-js.

Hope these experience can help you.

**Comment by jipyua** [comment]
After revisiting this post, the issue seems still exists, sometimes we failed to update the custom xml part, sometimes it fails to sync the state on UI, open an internal bug 9045585 to track.

**Comment by YbJerry** [comment]
> After revisiting this post, the issue seems still exists, sometimes we failed to update the custom xml part, sometimes it fails to sync the state on UI, open an internal bug 9045585 to track.

I think some customXmlPart like bibliography that we can only add it but can't edit it.

**Comment by jaiswalshubh1** [comment]
@YbJerry - apologies for how long this has been open, and thank you for the detail you put into reproducing it.

Where it actually stands: this is reproduced and understood. Updates to the custom XML part sometimes fail, and sometimes the change succeeds but the state shown in the UI does not follow it - which matches exactly what you demonstrated with the citation example. It was deprioritised as a long-standing platform gap rather than closed on merit.


cc: @kustonaut
